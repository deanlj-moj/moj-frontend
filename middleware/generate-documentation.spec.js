const { generateMarkdown } = require('./generate-documentation')

describe('generateMarkdown', () => {
  beforeAll(() => {
    jest.useFakeTimers('modern')
    jest.setSystemTime(new Date('2023-01-01T00:00:00Z'))
  })

  afterAll(() => {
    jest.useRealTimers()
  })

  it('should generate markdown with all fields populated', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Test Component',
        briefDescription: 'This is a brief description.',
        componentOverview: 'This is an overview.',
        componentProblemSolved: 'This is the problem solved.',
        howIsTheComponentUsed: 'This is how it is used.'
      },
      '/accessibility-findings-more': {
        accessibilityTellUsMore: 'Accessibility details here.'
      },
      '/prototype-url': {
        prototypeUrl: 'www.test.com'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Test User',
        emailAddress: 'test@test.com'
      }
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })

  it('should handle missing details gracefully', () => {
    const mockData = {
      '/component-details': {},
      '/component-code-details': {},
      '/additional-information': {},
      '/your-details': {}
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })

  it('should handle missing image data gracefully', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Another Component',
        briefDescription: 'Another brief description.',
        componentOverview: 'Another overview.',
        componentProblemSolved: 'Another problem solved.',
        howIsTheComponentUsed: 'Another usage.'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Another User',
        emailAddress: 'another@test.com'
      }
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })

  it('should sanitize the filename correctly', () => {
    const mockData = {
      '/component-details': {
        componentName: 'Component! With@ Special#Chars',
        briefDescription: 'Brief description here.',
        componentOverview: 'Overview here.',
        componentProblemSolved: 'Problem solved here.',
        howIsTheComponentUsed: 'Usage here.'
      },
      '/component-code-details': {
        componentCode: '<p>hello!</p>',
        howIsTheComponentUsed: 'Somehow'
      },
      '/additional-information': {
        additionalInformation: 'Additional information here.'
      },
      '/your-details': {
        fullName: 'Special User',
        emailAddress: 'special@test.com'
      }
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })

  it('should return minimal markdown if no valid data is provided', () => {
    const mockData = {
      '/component-details': {},
      '/component-code-details': {},
      '/additional-information': {},
      '/your-details': {}
    }

    const result = generateMarkdown(mockData)
    expect(result).toMatchSnapshot()
  })
})
