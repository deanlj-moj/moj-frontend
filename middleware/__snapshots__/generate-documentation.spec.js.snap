// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`generateMarkdown should generate markdown with all fields populated 1`] = `
{
  "content": "---
layout: layouts/component.njk
title: Test Component
type: component
eleventyNavigation:
  key: Test Component
  parent: Components
  excerpt: "This is a brief description."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

This is an overview.

## Why is the component needed?

This is the problem solved.

## Current uses for the component

This is how it is used.

{% endtab %}

{% tab "Code Stuff" %}

## Code Stuff

### Component Code


{% endtab %}

{% tab "Additional Info" %}

## Additional Info

Additional information here.

## Links

### Prototype


## Thing to consider

{% endtab %}

{% tab "Accessibility" %}

## Accessibility

Accessibility details here.

## Links

<a href="" target="_blank" rel="noopener noreferrer">Prototype example (opens in a new tab)</a>

## Thing to consider

{% endtab %}

{% tab "Contribution" %}

## Contribution

Test User

test@test.com

## History

**Test User** 01 January 2023

Component added

{% endtab %}

{% endtabs %}
",
  "filename": "component/documentation/test-component.md",
}
`;

exports[`generateMarkdown should handle missing details gracefully 1`] = `
{
  "content": "---
layout: layouts/component.njk
title: unknown-component
type: component
eleventyNavigation:
  key: unknown-component
  parent: Components
  excerpt: ""
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview



## Why is the component needed?



## Current uses for the component



{% endtab %}

{% tab "Code Stuff" %}

## Code Stuff

### Component Code


{% endtab %}

{% tab "Additional Info" %}

## Additional Info



## Links

### Prototype


## Thing to consider

{% endtab %}

{% tab "Accessibility" %}

## Accessibility



## Links

<a href="" target="_blank" rel="noopener noreferrer">Prototype example (opens in a new tab)</a>

## Thing to consider

{% endtab %}

{% tab "Contribution" %}

## Contribution





## History

**** 01 January 2023

Component added

{% endtab %}

{% endtabs %}
",
  "filename": "component/documentation/unknown-component.md",
}
`;

exports[`generateMarkdown should handle missing image data gracefully 1`] = `
{
  "content": "---
layout: layouts/component.njk
title: Another Component
type: component
eleventyNavigation:
  key: Another Component
  parent: Components
  excerpt: "Another brief description."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

Another overview.

## Why is the component needed?

Another problem solved.

## Current uses for the component

Another usage.

{% endtab %}

{% tab "Code Stuff" %}

## Code Stuff

### Component Code


{% endtab %}

{% tab "Additional Info" %}

## Additional Info

Additional information here.

## Links

### Prototype


## Thing to consider

{% endtab %}

{% tab "Accessibility" %}

## Accessibility



## Links

<a href="" target="_blank" rel="noopener noreferrer">Prototype example (opens in a new tab)</a>

## Thing to consider

{% endtab %}

{% tab "Contribution" %}

## Contribution

Another User

another@test.com

## History

**Another User** 01 January 2023

Component added

{% endtab %}

{% endtabs %}
",
  "filename": "component/documentation/another-component.md",
}
`;

exports[`generateMarkdown should return minimal markdown if no valid data is provided 1`] = `
{
  "content": "---
layout: layouts/component.njk
title: unknown-component
type: component
eleventyNavigation:
  key: unknown-component
  parent: Components
  excerpt: ""
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview



## Why is the component needed?



## Current uses for the component



{% endtab %}

{% tab "Code Stuff" %}

## Code Stuff

### Component Code


{% endtab %}

{% tab "Additional Info" %}

## Additional Info



## Links

### Prototype


## Thing to consider

{% endtab %}

{% tab "Accessibility" %}

## Accessibility



## Links

<a href="" target="_blank" rel="noopener noreferrer">Prototype example (opens in a new tab)</a>

## Thing to consider

{% endtab %}

{% tab "Contribution" %}

## Contribution





## History

**** 01 January 2023

Component added

{% endtab %}

{% endtabs %}
",
  "filename": "component/documentation/unknown-component.md",
}
`;

exports[`generateMarkdown should sanitize the filename correctly 1`] = `
{
  "content": "---
layout: layouts/component.njk
title: Component! With@ Special#Chars
type: component
eleventyNavigation:
  key: Component! With@ Special#Chars
  parent: Components
  excerpt: "Brief description here."
---

{% tabs "Contents" %}

{% tab "Overview" %}

## Overview

Overview here.

## Why is the component needed?

Problem solved here.

## Current uses for the component

Usage here.

{% endtab %}

{% tab "Code Stuff" %}

## Code Stuff

### Component Code


{% endtab %}

{% tab "Additional Info" %}

## Additional Info

Additional information here.

## Links

### Prototype


## Thing to consider

{% endtab %}

{% tab "Accessibility" %}

## Accessibility



## Links

<a href="" target="_blank" rel="noopener noreferrer">Prototype example (opens in a new tab)</a>

## Thing to consider

{% endtab %}

{% tab "Contribution" %}

## Contribution

Special User

special@test.com

## History

**Special User** 01 January 2023

Component added

{% endtab %}

{% endtabs %}
",
  "filename": "component/documentation/component--with--special-chars.md",
}
`;
